docs: Add Wave api rate limit guide (#914)
seqeralabs/wave: docs/nextflow/reduce-api-calls.md

---
title: Reducing Wave API calls
description: Learn how to configure Nextflow to freeze containers and reduce API calls
date created: 2025-09-30
date edited: 2025-10-31
tags: [nextflow, wave, rate limits, guides]
---

Large-scale pipelines that pull container images across thousands of concurrent tasks can encounter Wave rate limits. This guide describes how to configure your Nextflow pipeline to use Wave freeze and reduce API calls to avoid rate limits.

:::note
Wave applies rate limits to container builds and pulls (manifest requests). Authenticated users have higher rate limits than anonymous users. See [API limits](../api.md#api-limits) for more information.
:::

## How Wave freeze avoids rate limits

> Wave freeze builds your container image once and stores it in your registry. After the initial build, the source of the container image manifest and layers are redirected to your private registry by Wave.

Reviewer: I Quite like the freeze description we have in the wave Q&A Draft google doc as it covers some additional value adds of using wave freeze which may remove the need to draw comparison between wave request flows (Build , Augment , Freeze & Mirror)

---
title: Reduce Wave API calls
description: Learn how to use Wave freeze to reduce API calls and avoid rate limits in large-scale Nextflow pipelines
date created: 2025-09-30
date edited: 2025-09-30
tags: [nextflow, wave, rate limits, guides]
---

Wave rate limits can affect large-scale pipelines that pull containers across thousands of concurrent tasks. This guide shows you how to use Wave freeze to reduce API calls and avoid rate limits.

## API limits

Wave applies rate limits to container builds and pulls. Authenticated users have higher rate limits than anonymous users.

If an access token is provided, the following rate limits apply:

- 250 container builds per hour

> - 2,000 container pulls per minute

Author: Added 👍

> - 2,000 container pulls (manifests) per minute

Author: ```suggestion
- 2,000 container pulls (manifest requests) per minute
```

If an access token isn't provided, the following rate limits apply:

- 25 container builds per day

> - 100 container pulls per hour

Author: ```suggestion
- 100 container pulls (manifests) per hour
```

## How Wave pull rate limits work

When you pull a container:

- The manifest request counts as one pull against your rate limit
- Layer and blob requests don't count against rate limits
- A container with 100 layers counts as 1 pull

Rate limits affect pipelines with high concurrency. The following example demonstrates this issue:

- 50 concurrent pipeline runs
- Each run spawns 10,000 tasks
- Each task pulls a container
- Total: 500,000 manifest requests

This volume exceeds the 2,000 per minute limit and causes failed tasks and pipeline errors.

## Use Wave freeze to avoid rate limits

Wave freeze builds your container once and stores it in your registry. After the initial build, all container pulls bypass Wave.

**Building without freeze:**

1. Each task requests a manifest from Wave

> 1. Wave retrieves the base image from the source registry

Author: Good shout.

1. Wave builds the image with the fusion layer
1. Wave returns the modified manifest
1. Every task creates one API call to Wave

With thousands of concurrent tasks, this approach exceeds rate limits.

> **Building with freeze:**

Author: I've chatted to @munishchouhan and confirmed this is correct.

I've added a reminder that frozen images will only be reused as long as the image remains the same, and a new section to recommend using specific version tags or SHA256 digests with Wave freeze.

1. Wave builds the image once with your specifications
1. Wave pushes the complete image to your registry
1. Wave returns a direct URL to your registry
1. All future pulls go directly to your registry

With freeze enabled, Wave is removed from the container pull path. Your compute instances pull directly from your registry with no Wave API calls.

> ## Configure Wave freeze

Author: Later on we talk about the Fusion layer. I have included it in the config, commented it as recommended, and added a note to say why it's recommended but not required.

To configure Wave freeze, add the following configuration to your Nextflow pipeline:

```groovy
wave.enabled = true
wave.freeze = true
wave.build.repository = '<BUILD_REPOSITORY>'

tower.accessToken = '<TOWER_ACCESS_TOKEN>'
```

Replace the following:

- `<BUILD_REPOSITORY>`: your repository to store and access built containers
- `<TOWER_ACCESS_TOKEN>`: your Seqera access token

### Container registry selection

**Recommended**: Use Amazon ECR for AWS Batch workloads

Amazon ECR has the following benefits:

- Automatic authentication through IAM roles
- No manual credential configuration
- Lowest latency for AWS workloads
- Simplest setup

> **Not recommended**: Private Docker Hub for AWS Batch workloads

Author: I've softened to "**Not recommended**: External container registries for AWS Batch workloads" as per your other suggestion.

Are you suggesting we don't specifically call out AWS? And just refer to "native cloud container registries" instead of "AWS Batch workloads" for both recommended and not recommended?

Private Docker Hub has the following limitations:

- Requires manual credential configuration on each compute instance
- Additional security overhead
- More complex authentication setup

If you use private Docker Hub, configure Docker credentials on each compute instance.

### First pipeline run

When you run your pipeline for the first time with Wave freeze:

1. The Nextflow head job sends your build request to Wave
1. Wave checks whether the requested images already exist
1. Wave builds any missing images and pushes them to your registry
1. Wave returns the final registry URLs
1. Your compute tasks pull images directly from your registry

The initial build counts against build limits, not pull limits

### Subsequent pipeline runs

When you run your pipeline again:

1. The Nextflow head job contacts Wave to check for existing images
1. Wave finds the cached images (matched by content hash)
1. Wave returns the registry URLs immediately without rebuilding
1. All container pulls go directly to your registry
1. No Wave API calls occur during task execution

> 

Reviewer 2: This is partially correct, the pull limit applies also when requesting the freeze container, tho same container will serve many job requests 

https://github.com/seqeralabs/wave/blob/479f50ef9dfd3a630811f6c3dcb0bbe5ad79d280/src/main/groovy/io/seqera/wave/controller/ContainerController.groovy#L266-L267

Author: Thanks Paolo, folding it into changes now.

On subsequent attempts to access the container, Wave will not rebuild the existing container:

1. Wave checks for existence of the container
1. Wave identifies the container exists at the Wave Freeze configured location
1. Wave returns the existing URL to your registry

With freeze enabled, Wave is removed from the container pull path. Your compute instances pull directly from your registry with no Wave API calls.

## Configure Wave freeze

To configure Wave freeze, add the following configuration to your Nextflow pipeline:

```groovy
wave.enabled = true
wave.freeze = true
wave.build.repository = '<BUILD_REPOSITORY>'

tower.accessToken = '<TOWER_ACCESS_TOKEN>'
```

Replace the following:

- `<BUILD_REPOSITORY>`: your repository to store and access built containers
- `<TOWER_ACCESS_TOKEN>`: your Seqera access token

### Container registry selection

**Recommended**: Use Amazon ECR for AWS Batch workloads

Amazon ECR has the following benefits:

- Automatic authentication through IAM roles
- No manual credential configuration
- Lowest latency for AWS workloads
- Simplest setup

**Not recommended**: Private Docker Hub for AWS Batch workloads

> Private Docker Hub has the following limitations:

Author: ```suggestion
External container registries have following limitations:
```

- Requires manual credential configuration on each compute instance
- Additional security overhead
- More complex authentication setup

> If you use private Docker Hub, configure Docker credentials on each compute instance.

Author: ```suggestion
If you use external container registries, configure your credentials on each compute instance.
```

---

> title: Reducing Wave API calls

Reviewer: I would change this title to something like  `Wave Freeze` and have the reduction of API calls be a secondary effect.

The benefit of wave freeze is enhanced security , availability and reliability by reducing the need for wave as part of a pipeline build, acting as a request proxy.

---
title: Reduce Wave API calls
description: Learn how to freeze containers and reduce API calls in large-scale Nextflow pipelines
date created: 2025-09-30
date edited: 2025-10-01
tags: [nextflow, wave, rate limits, guides]
---

Wave rate limits can affect large-scale pipelines that pull container images across thousands of concurrent tasks. This guide shows you how to use Wave freeze to reduce API calls and avoid rate limits.

## API limits

Wave applies rate limits to container builds and pulls (manifest requests). Authenticated users have higher rate limits than anonymous users.

If an access token is provided, the following rate limits apply:

- 250 container builds per hour
- 2,000 container pulls (manifest requests) per minute

If an access token isn't provided, the following rate limits apply:

- 25 container builds per day
- 100 container pulls (manifest requests) per hour

## How Wave pull rate limits work

> When you pull a container image:

Author: ```suggestion
When you pull a container image:

1. The Docker service downloads the container manifest (a JSON file detailing each layer of the final image).
1. The Docker service downloads each layer listed in the manifest file.

Wave defines a pull as downloading both the container manifest and all layers. Therefore:
```

1. The Docker service downloads each layer listed in the manifest file.

Wave defines a pull as downloading both the container manifest and all layers. Therefore:

- The manifest request to Wave counts as one pull against your rate limit
- Layer and blob requests don't count against rate limits
- A container image with 100 layers counts as 1 pull

Rate limits affect pipelines with high concurrency. The following example demonstrates this issue:

- 50 concurrent pipeline runs
- Each run spawns 10,000 tasks
- Each task pulls a container image
- 500,000 manifest requests are made

This volume exceeds the 2,000 container pulls per minute limit and causes failed tasks and pipeline errors.

## How Wave freeze avoids rate limits

Wave freeze builds your container image once and stores it in your registry. After the initial build, the source of the container image manifest and layers are redirected to your private registry by Wave.

### Building without Wave freeze

When you run your pipeline without Wave freeze:

1. Each task requests a manifest from Wave.
1. Wave performs one of the following actions:

>     - Retrieves the base image manifest from the source registry

Reviewer 3: I'm not sure I fully understand this first bullet. Doesn't building entail having to add extra packages to a base image?  If base image only is being pulled and returned this seems more like the augmentation flow?

    - Builds the image from a Dockerfile
    - Builds the image from a Conda definition
1. Wave injects the Fusion layer to the container image manifest.
1. Wave stores the final manifest on Seqera infrastructure.
1. Wave returns the modified manifest.

> 1. Every task creates one API call to Wave.

Reviewer 3: I would lose this bullet -- you have a good description of how we charge API calls above, and this now mixes build mechanics with how we rate-limit.

This approach exceeds rate limits with thousands of concurrent tasks.

### Building with Wave freeze

When you run your pipeline with Wave freeze for the first time:

1. The Nextflow head job sends your build request to Wave.

> 1. Wave checks whether the requested images already exist.

Author: Can I have some more context here please? Is there a recording or other thread I can review?

@robsyme @gavinelder @gwright99

1. Wave builds any missing images and pushes the manifest and layers to your registry.
1. Wave returns the final registry URLs.
1. Your compute tasks pull images directly from your registry.

When you run your pipeline with Wave freeze again:

1. The Nextflow head job contacts Wave to request the frozen images.
1. Wave finds the frozen images in your registry (matched by content hash).
1. Wave returns the container URLs in the destination container registry without rebuilding.
1. All tasks pull the image directly from your registry.

> 1. The frozen image serves many task executions.

Author: Yup - good call.

The cache reuses unchanged layers to reduce build times and costs.
:::

:::note
You can use Wave freeze without Fusion.
However, it ensures that the frozen images contain the necessary components for optimal performance when accessing cloud object storage.
:::

## Container image tags

Use specific version tags (such as `ubuntu:22.04`) or SHA256 digests with Wave freeze.
Specific tags enable Wave to match content hashes and reuse frozen images.
This ensures reproducibility and eliminates unnecessary rebuilds.
Avoid using the `latest` tag because it points to different image versions over time.

## Container registry selection

**Recommended**: Use your cloud provider's native container registry for the simplest setup and integration.

Native cloud registries have the following benefits:

- Automatic authentication through cloud IAM roles
- Low latency for workloads in the same cloud region
- Simple setup and configuration
- Native integration with your cloud platform

Examples of native registries by cloud provider:

- **AWS**: Amazon Elastic Container Registry (ECR)
- **Azure**: Azure Container Registry (ACR)
- **Google Cloud**: Google Artifact Registry

> **Not recommended**: Third-party container registries.

Author: I'll keep the text here minimalist and reframe as "alternate", and add text to highlight the additional effort and public endpoints.

---
title: Reduce Wave API calls
description: Learn how to use Wave freeze to reduce API calls and avoid rate limits in large-scale Nextflow pipelines
date created: 2025-09-30
date edited: 2025-09-30
tags: [nextflow, wave, rate limits, use cases, guides]
---

Wave rate limits can affect large-scale pipelines that pull containers across thousands of concurrent tasks. This guide shows you how to use Wave freeze to reduce API calls and avoid rate limits.

## How Wave rate limits work

> Wave applies rate limits to manifest requests, not layer requests. With an access token, you can pull 2,000 containers per minute. When you pull a container:

Reviewer 4: I think we only need to say, that Wave rate limit applies as per container pull and not each container layer. no need to specify manifest.

Author: Confirming I've changed this to "Wave applies rate limits to container pulls."

Author: Changed to "...pulls (manifest requests)..."
